### DISCUSSION TITLE
Nested transaction hangs when used concurrently

### DISCUSSION THREAD

**Discussion by ljiatu** [sqlalchemy/sqlalchemy#5896] [Usage Questions] [answered]
**Describe your question**

**Example - please use the [Minimal, Complete, and Verifiable](http://stackoverflow.com/help/mcve) guidelines if possible**
This is a noob question. I'm trying to test how two concurrent transactions with nested transactions behave and seem to run into some sort of deadlock. Here's what I have:

```
@pytest.mark.asyncio
async def test_nested_transactions(
        write_through_db_session: Session, db_engine: Engine, db_account: Account
) -> None:
    write_through_db_session.add(db_account)
    account_id = db_account.id
    write_through_db_session.commit()

    @async_wrapper
    @nested_transaction_wrapper
    def update_account_balance(session: Session) -> None:
        time.sleep(2)
        print("-" * 80)
        print(f"{id(session)} querying!")
        print("-" * 80)
        account = session.query(Account).get(account_id)
        account.balance += 10
        print("-" * 80)
        print(f"{id(session)} updated balance!")
        print("-" * 80)

    session1 = create_db_session(db_engine)()
    session2 = create_db_session(db_engine)()

    try:
        await asyncio.gather(
            update_account_balance(session1),
            update_account_balance(session2),
        )
        session1.commit()
        session2.commit()
    finally:
        session1.close()
        session2.close()
```

`async_wrapper` is defined as below:
```
def async_wrapper(f: Callable):
    @functools.wraps(f)
    async def wrapper(*args, **kwargs) -> Any:
        event_loop = asyncio.get_event_loop()
        return await event_loop.run_in_executor(None, functools.partial(f, *args, **kwargs))


    return wrapper
```

`nested_transaction_wrapper` is defined as below:
```
def nested_transaction_wrapper(f: Callable) -> Callable:
    @wraps(f)
    def wrapper(session: Session, *args, **kwargs) -> Any:
        session.begin_nested()

        try:
            print("-" * 80)
            print(f"{id(session)} beginning!")
            print("-" * 80)
            result = f(session, *args, **kwargs)
            print("-" * 80)
            print(f"{id(session)} committing!")
            print("-" * 80)
            session.commit()
            return result
        except Exception:
            session.rollback()
            raise

    return wrapper
```

`write_through_db_session` is just a regular DB session that commits to the underlying test DB. `db_account` is a fake `Account` object I created for testing.

I added a bunch of print statements to help with debugging. Output below:
```
--------------------------------------------------------------------------------
4578504768 beginning!
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
4578506256 beginning!
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
4578504768 querying!
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
4578506256 querying!
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
4578504768 updated balance!
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
4578504768 committing!
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
4578506256 updated balance!
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
4578506256 committing!
--------------------------------------------------------------------------------
```

The program hangs after the last message is printed out, which means the second session can't commit the nested transaction. I don't quite understand why the second transaction would be blocked since it's just trying to commit its own nested transaction, not even the outer transaction yet. Am I completely misunderstanding how nested transactions work?

I'm also not quite sure whether this is a SQLAlchemy question or a Postgres SAVEPOINT question. My apologies in advance if I asked in the wrong place.

Thank you so much for your help, and for your amazing work on SQLAlchemy!

**Complete stack trace, if applicable**
N/A. The program hangs.

**Versions**
 - OS: maxOS Catalina 10.15.6
 - Python: 3.8.5
 - SQLAlchemy: 1.3.22
 - Database: Postgres 12.3
 - DBAPI: psycopg2 2.8.6

**Additional context**
N/A.

**Useful links**
- [Get Support](https://www.sqlalchemy.org/support.html) on the website
- The [documentation](https://docs.sqlalchemy.org/en/latest/) website
- The [UsageRecipes](https://github.com/sqlalchemy/sqlalchemy/wiki/UsageRecipes) wiki
- [Stack Overflow](https://stackoverflow.com/questions/tagged/sqlalchemy) tag
- SQLAlchemy [Google group](http://groups.google.com/group/sqlalchemy)
- [Gitter](https://gitter.im/sqlalchemy/community) chat

**Have a nice day!**

**Comment by zzzeek** [comment, marked as answer]
So I don't think the asyncio aspect of this is helping , as it is quite unusual to use asyncio in such a way as everything that's being invoked is using blocking IO; I see that part of the operation is using a run_in_executor() to try to run things in a thread,  but If you just want to run two transactions with blocking IO in two threads, it's a whole lot less code to just use a Thread, just like this:

```python


threads = [
    threading.Thread(target=function, args=(session, )),
    threading.Thread(target=function, args=(session, ))
]
for t in threads:
    t.start()
for t in threads:
    t.join()

```

that said, if you want to work with asyncio, then you will have more idiotmatic code if you [use the asyncio interface](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html), which works great, then you dont have to deal with run_in_executor().

next part, where you want to see what's going on.  much better way than print statements is to set echo=True or echo='debug' on your create_engine().    although it won't out of the box give you an easy way to distinguish between which transaction is which, you'd probably have to make a custom logger that uses some context for that, but in this case to figure out what's going wrong you could just write everything above serially and step through it, since it is pretty much doing most of the stuff serially in any case above.

the next thing to know is, even if you aren't printing out the SQL, whenever your PostgreSQL program deadlocks, go to the psql command line and just look for the transactions:

```
test=> select wait_event_type, query from pg_stat_activity where usename='scott';
 wait_event_type |                                   query                                    
-----------------+----------------------------------------------------------------------------
 Client          | RELEASE SAVEPOINT sa_savepoint_1
 Lock            | UPDATE a SET x=15 WHERE a.id = 1
                 | select wait_event_type, query from pg_stat_activity where usename='scott';
(3 rows)

```

you can see deadlocks for small experiements quickly in the unix ps listing too, since PostgreSQL uses a process for every connection:

```
$ ps -ef | grep "postgres: scott"
postgres 1936661   22962  0 Jan29 ?        00:00:00 postgres: scott test ::1(38540) idle in transaction
postgres 1936662   22962  0 Jan29 ?        00:00:00 postgres: scott test ::1(38542) UPDATE waiting
classic  1937371 1826216  0 00:00 pts/1    00:00:00 grep --color=auto postgres: scott

```

the next thing is, having that session.commit() , as well as the session.close(), outside of your "run in executor" is wrong from a concurrency point of view, since session.commit() and session.close() are blocking.    In fact in commit() is normally where you'd have the whole flush take place, that's where your UPDATE statements and everything would be happening if you didnt have that savepoint commit happening first.   They still can run and complete, if you didn't have the dealdock also happening, because asyncio will run your blocking IO without any  issue but it will block all other asyncio things from happening, which negates the entire point of using asyncio.

Additionally,  in this case the actual deadlock is because the commit() for the first transaction can't complete because it is waiting for the second transaction to emit its UPDATE, so moving it into the executor block would fix the problem here.

the "deadlock" can be seen like this without any async or threads:

```python
s1 = Session(e)
s2 = Session(e)


a1 = s1.query(A).get(1)
a2 = s2.query(A).get(1)


a1.x += 5
a2.x += 10

s1.flush()
s2.flush()

s1.commit()
s2.commit()


```

the SQLAlchemy Session in your example is emitting a flush() when those savepoints end.   The deadlock is that transaction A has emitted an UPDATE on the row so has a lock on it (specifically an [exclusive row level lock](https://www.postgresql.org/docs/9.1/explicit-locking.html)), transaction  B then also emits an UPDATE on the row so attempts to get a lock on it, and has to wait for A to finish its transaction before it can do so.   but your program isn't committing the transaction for A before B finishes, so nothing can complete.

in fact "13.3.3. Deadlocks" in that link https://www.postgresql.org/docs/9.1/explicit-locking.html illustrates the identical UPDATE queries you are running here as their example of how to make a deadlock.

you can fix your program by making sure the outer .commit() also happens inside the run_in_executor part.

**Reply by ljiatu** [reply]
Thanks for the fast and detailed response as always @zzzeek! Your explanation makes a lot of sense and I can see now where things are going wrong.

So re: `asyncio` interface - we are on SQLAlchemy 1.3 right now, and we haven't started the 1.4 migration yet. I'm waiting for 1.4.0b2 to come out before starting the migration since you mentioned in an earlier discussion that b2 fixes some of the big problems in b1. The fact that the doc mentions the `async` support is "alpha level" (https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html) also makes me a bit antsy.

Re: print vs `echo=True`. I actually started with `echo=True`, but then tried to use prints because of exactly the reason you said, that `echo=True` doesn't give me concrete information on which transaction is doing all the stuff, hence I decided just to print out the session ids to make things clearer.

Re: `commit()` and `close()` are blocking - you are absolutely right. That was a miss on my part. I should have those made async as well, although it wouldn't have helped in this situation.

The reason we have `nested_transaction_wrapper` is to break large functions that manipulate different kinds of objects in the DB into smaller, reusable pieces that manipulate one kind of objects at a time. To achieve atomicity in the large function, we want to wrap each piece into a nested transaction, so that the large function only commits at the very end if all the smaller pieces go through. If any of the smaller piece fails, we can just rollback the outer transaction and nothing gets committed. We are also explicitly grabbing locks using `with_for_update` on the objects we are manipulating to prevent race conditions, so the `rollback` in `nested_transaction_wrapper` will immediately release the lock held in the wrapped function if anything goes wrong so that we don't hold onto locks longer than necessary. The other thing is that since `nested_transaction_wrapper` rolls back the session on error, the caller can start with a clean slate and do its own operations w/o having to explicitly rollback. 

The problem is, we are not confident that we are always grabbing locks in a consistent order, which means we may end up getting frequent deadlocks. So now we are trying to use the "SERIALIZABLE" isolation level so that we don't need to explicitly grab locks, and this can potentially simplify future development quite a bit. I can see us getting rid of the `nested_transaction_wrapper`s, which gets rid of all the flushes until the `commit()` at the very end of the large function. Would you say this is the right way to go?

If you have any other suggestions to achieve what we are trying to do, I'm all ears. Thanks again for your help!

**Reply by ljiatu** [reply]
Sorry if this sounded abstract, here are some pseudocode examples. Let's say we have a function that handles the transaction on a listing in a marketplace application. The two accounts involved are `buyer_account` and `seller_account`. If we put everything into one function, then it would look like this (minus `None` checks for simplicity):

```
def transact(session: Session, listing_id: UUID, buyer_account_id: UUID, seller_account_id: UUID) -> None:
    listing = session.query(Listing).filter(Listing.id == listing_id).with_for_update().populate_existing().one()
    # Some state checks to make sure the listing is transact-able.
    ...

    accounts = session.query(Account).filter(Accout.id in [buyer_account_id, seller_account_id]).with_for_update().populate_existing().all()
    # Assign buyer_account and seller_account based on ID from the query result.
    # Make sure the buyer account has enough balance, and update the balances for the two accounts.
    ...

    # Update the listing state and the sold item's status.
    item = session.query(Item).filter(Item.listing_id = listing_id).with_for_update().populate_existing().one()
    listing.state = "SOLD"

    # Update the buyer's and seller's account statements to record the transaction.
    ...
```

The function is roughly 300 hundred lines if written out, and updates different kinds of objects at once. There are also parts of the function that are re-usable by other functions. So we decided to break it up into smaller functions. Since we want to maintain atomicity (i.e. if updating one object fails then fail the entire operation), we decided to wrap the smaller pieces into nested transactions using `nested_transaction_wrapper`. So the functions look like below now:

```
@nested_transaction_wrapper
def get_listing_for_update(session: Session, listing_id: UUID) -> Listing:
    listing = session.query(Listing).filter(Listing.id == listing_id).with_for_update().populate_existing().one()
    # Some state checks to make sure the listing is transact-able.
    ...

@nested_transaction_wrapper
def update_account_balance(session: Session, buyer_account_id: UUID, seller_account_id: UUID) -> Tuple[Account, Account]:
    accounts = session.query(Account).filter(Accout.id in [buyer_account_id, seller_account_id]).with_for_update().populate_existing().all()
    # Assign buyer_account and seller_account based on ID from the query result.
    # Make sure the buyer account has enough balance, and update the balances for the two accounts.
    ...

@nestes_transaction_wrapper
def create_transaction_records(buyer_account: Account, seller_account: Account) -> None:
    # Update the buyer's and seller's account statements to record the transaction.
    ...

def transact(session: Session, listing_id: UUID, buyer_account_id: UUID, seller_account_id: UUID) -> None:
    listing = get_listing_for_update(session, listing_id)
    buyer_account, seller_account = update_account_balance(session, buyer_account_id, seller_account_id)
    
    # Update the listing state and the sold item's status.
    item = session.query(Item).filter(Item.listing_id = listing_id).with_for_update().populate_existing().one()
    listing.state = "SOLD"

    create_transaction_records(buyer_account, seller_account)

    # Commits the entire transaction
    session.commit()
```

So our goal is to keep the code modular, maintain atomicity, and use the "SERIALIZABLE" isolation level.

Hope this makes what I was trying to say clearer! Thanks again!

**Reply by zzzeek** [reply]
I've never heard of SERIALIZABLE isolation used to *reduce* deadlocks, it's going to increase them and it's also going to reduce performance and concurrency, possibly to a significant degree.  PostgreSQL's default isolation of "read committed" is very low and you really shouldn't have problems with deadlocks at the database level.    the deadlock illustrated at the top of your issue wasn't as much a database deadlock as it was a deadlock in your application.

some background on SERIALIZABLE: https://stackoverflow.com/questions/39029573/why-does-a-serializable-isolation-level-lead-to-deadlock-and-concurrency-issues

also http://shiroyasha.io/transaction-isolation-levels-in-postgresql.html

> This guarantee comes with a price. Serializable transaction error out with Serializable issues frequently, and there is an additional performance cost to be paid. I would advise using Serializable transaction only if you have a deep understanding of the PostgreSQL engine.

I would agree, I would never personally go near SERIALIZABLE isolation and I've yet to see an issue where there is not a much better way to solve the problem.

**Reply by zzzeek** [reply]
another great post on this, some background on PG's serializable level, a shop that went all serializable and then had to move down to "repeatable read" (RR is a pretty useful isolation level IMO):   https://peter.grman.at/postgres-repeatable-read-vs-serializable/

**Reply by ljiatu** [reply]
Thanks for the links! I also came across the last article a while ago where they switched from SERIALIZABLE to REPEATABLE READ, and we were thinking that we could potentially also use SERIALIZABLE to start with, retry serialization failures transparently (to the user), not worry about explicit locking, and monitor how often serializable failure happens and adjust if needed. The deadlock I'm concerned about is not the kind I had in the example above - that was just me testing how would nested transactions behave with different transaction isolation levels. I'm more concerned about ensuring consistent locking order in different parts of the application, which could get harder to track as the application grows. But again, it could be a misguided concern.

Re: SERIALIZABLE isolation level increasing deadlocks, why so? I was reading the Postgres documentation and it mentions this (https://www.postgresql.org/docs/current/transaction-iso.html):

> To guarantee true serializability PostgreSQL uses predicate locking, which means that it keeps locks which allow it to determine when a write would have had an impact on the result of a previous read from a concurrent transaction, had it run first. In PostgreSQL these locks do not cause any blocking and therefore can not play any part in causing a deadlock.

Would it cause a different kind of deadlock?

**Reply by ljiatu** [reply]
Ah I see where deadlocks can still happen. They could still come from inconsistent order of updates, since the DB will still grab row-level exclusive locks. The paragraph I quoted above only says that the introduction of predicate locks won't cause deadlocks on their own.

I'm still curious though as to why SERIALIZABLE isolation would _increase_ deadlocks. Is it because of not carefully ordering updates due to incorrectly thinking that SERIALIZABLE isolation can help eliminate them?

**Reply by zzzeek** [reply]
for PG, SERIALIZABLE might lead to rejected transactions more than deadlocks.   PG doesn't actually "deadlock" too much on its own, the "deadlocks" we see with PG are what we have above, where the application itself has multiple transactions that themselves are in the same thread.      SERIALIZABLE will lead to errors because it's more likely two transactions can't be reconciled serially.    it also would reduce tps quite a bit if that's something that's important.

**Reply by ljiatu** [reply]
Ok that makes sense. Thanks for all your help!

**Reply by zzzeek** [reply]
> 
> Re: print vs `echo=True`. I actually started with `echo=True`, but then tried to use prints because of exactly the reason you said, that `echo=True` doesn't give me concrete information on which transaction is doing all the stuff, hence I decided just to print out the session ids to make things clearer.

I've added #5911 which I'm going to implement now for 1.4.0.b2 because I was surprised how hard this is to do.
